### ISSUE TITLE
The ability to store passwords using different hashing algorithms is limited

### ISSUE THREAD

**Issue by maderlock** [magento/magento2#5463]
## Precondtions
Magento `develop` branch

## Steps to reproduce
1. Create passwords for users hashed using different hashing algorithms (e.g. md5 and sha256)
2. Attempt to login against these
## Expected result
1. Both would log in
## Actual result

1.Those created with sha256 work
2. Those created with anything else (e.g. md5) fail incorrectly

The reason for this is that although the version of the hashing algorithm is stored against the hashed password, this is not used for creating the hash for comparison, which always defaults to sha256.

This could be fixed by changing the following line in Magento\Framework\Encryption\Encryptor::isValidHash from:

```
return implode(
        self::DELIMITER,
        [
            $this->hash($salted),
            $salt,
            $version
        ]
    );
```

to:

```
return implode(
        self::DELIMITER,
        [
            $this->hash($salted, $version),
            $salt,
            $version
        ]
    );
```

There are also various limitations to adding other hashing algorithms (a common task when migrating users from non-Magento systems), primarily that there are so many private methods in this class! Please make them protected so that we don't have to copy and paste them when overriding.

**Comment by magento-engcom-team** [comment]
@maderlock, thank you for your report.
The issue is already fixed in 2.1.10

**Comment by k4emic** [comment]
This issue was **not** solved in 2.1.10 and is still present in the development branches for 2.1, 2.2 and 2.3.

2.1-develop: https://github.com/magento/magento2/blob/2.1.10/lib/internal/Magento/Framework/Encryption/Encryptor.php#L155

2.2-develop: https://github.com/magento/magento2/blob/2.2-develop/lib/internal/Magento/Framework/Encryption/Encryptor.php#L155

2.3-develop: https://github.com/magento/magento2/blob/2.3-develop/lib/internal/Magento/Framework/Encryption/Encryptor.php#L160

The following testcase replicates the problem:

```php
use Magento\Framework\App\DeploymentConfig;
use Magento\Framework\Encryption\Encryptor;
use Magento\Framework\Math\Random;

class PasswordTest extends \PHPUnit\Framework\TestCase
{

    protected $random;
    protected $deploymentConfig;
    protected $encryptor;

    public function setUp()
    {

        $this->random = $this->createMock(Random::class);
        $this->deploymentConfig = $this->createMock(DeploymentConfig::class);

        $this->encryptor = new Encryptor(
            $this->random,
            $this->deploymentConfig
        );
    }

    public function testHashingShouldUseSpecifiedAlgo()
    {
        $password = 'password123';
        $salt = 'random_salt';

        $sha256Hash = $this->encryptor->getHash($password, $salt, Encryptor::HASH_VERSION_SHA256);
        $this->assertTrue($this->encryptor->isValidHash($password, $sha256Hash));

        $md5Hash = $this->encryptor->getHash($password, $salt, Encryptor::HASH_VERSION_MD5);
        $this->assertTrue($this->encryptor->isValidHash($password, $md5Hash));

        list($hashMd5) = explode(Encryptor::DELIMITER, $md5Hash);
        list($hashSha256) = explode(Encryptor::DELIMITER, $sha256Hash);

        $this->assertNotEquals($hashMd5, $hashSha256);
    }
}
```

The hash strings are identical, because sha256 is used for both hashing operations.

**Comment by k4emic** [comment]
@magento-engcom-team Please reopen this issue.

**Comment by magento-engcom-team** [comment]
Hi @maderlock. Thank you for your report. 
The issue has been fixed in magento/magento2#13884 by @k4emic in 2.2-develop branch
Related commit(s):
 - [adab3e55ea7d0e6406e16d5798ce7ff5f230b7fe](https://github.com/magento/magento2/commit/adab3e55ea7d0e6406e16d5798ce7ff5f230b7fe)

The fix will be available with the upcoming 2.2.5 release.

**Comment by magento-engcom-team** [comment]
Hi @maderlock. Thank you for your report. 
The issue has been fixed in magento/magento2#13886 by @k4emic in 2.1-develop branch
Related commit(s):
 - [86dd6ecf16199fed13ab030e9794160dd3a37f2e](https://github.com/magento/magento2/commit/86dd6ecf16199fed13ab030e9794160dd3a37f2e)

The fix will be available with the upcoming 2.1.14 release.

**Comment by magento-engcom-team** [comment]
Hi @maderlock. Thank you for your report. 
The issue has been fixed in magento/magento2#13885 by @k4emic in 2.3-develop branch
Related commit(s):
 - [d537d56e440079a7db52419267d18220b23241e5](https://github.com/magento/magento2/commit/d537d56e440079a7db52419267d18220b23241e5)

The fix will be available with the upcoming 2.3.0 release.
